### PR TITLE
[Core] Non-iterative local coords Triangle3D3N

### PR DESCRIPTION
**Pull request by loumalouomega**
This uses the analytic solution of the local coordinates for the case of the Triangle3D3N to avoid iterative approach

@KratosMultiphysics/technical-committee @KratosMultiphysics/implementation-committee

### CODE REVIEW

File: kratos/geometries/triangle_3d_3.h [hunk 1]
```diff
@@ -879,7 +875,66 @@ template<class TPointType> class Triangle3D3
...
+        if (std::abs(distance) > std::numeric_limits<double>::epsilon()) { // Not in the plane, projecting
+            noalias(point_projected) = rPoint - normal * distance;
+        }
+
+        // Computation of the rotation matrix
+        BoundedMatrix<double, 3, 3> rotation_matrix = ZeroMatrix(3, 3);
+        for (IndexType i = 0; i < 3; ++i) {
+            rotation_matrix(0, i) = tangent_xi[i];
+            rotation_matrix(1, i) = tangent_eta[i];
+        }
```
Comment (jcotela) [inline kratos/geometries/triangle_3d_3.h:905]: Don't we need the third row to be the normal[i] here?

Comment (loumalouomega) [reply kratos/geometries/triangle_3d_3.h:905]: This is the algorithm I use in the mortar segmentation, I think is not needed:

https://github.com/KratosMultiphysics/Kratos/blob/2ef11c5c6baebc205d4e25606254a3059a14d18b/kratos/utilities/mortar_utilities.h#L185

Comment (jcotela) [reply kratos/geometries/triangle_3d_3.h:905]: Ok, you know your algorithms… Just saying that this is not a rotation matrix then :P

Comment (loumalouomega) [reply kratos/geometries/triangle_3d_3.h:905]: True, projection to base matrix?

**Comment by loumalouomega** [comment]
@frastellini I think this can be a way to improve performance in the explicit (and implicit too BTW) contact, because the local coordinates can be done in a non-iterative way

**Comment by loumalouomega** [comment]
I run all the tests and is working, I think this is a must for increasing performance in some applications

**Review by frastellini — APPROVED** [review]
Seems ok to me

**Comment by loumalouomega** [comment]
Like this is important I will ask for anyone from @KratosMultiphysics/technical-committee to say OK

**Comment by philbucher** [comment]
did you run the tests in the MappingApp?

I think you broke sth, at least the test was working before

**Comment by loumalouomega** [comment]
No I did not compiled the mapping. Sorry if I broke something.

**Comment by philbucher** [comment]
happens to the best :) 
Are you using this function in your stuff? If so did you have to change sth?

**Comment by loumalouomega** [comment]
Yes, I use it intensively. The tests in the core/fluid/structural/contactstructural are working. Can you tell me which is the failing test?, maybe previously it reached a convergence in the jacobian residual but the point was actually outside

**Comment by philbucher** [comment]
Hm strange, then it might be a corner-case :/

I will check it and let you know if I need help

File: kratos/geometries/triangle_3d_3.h [hunk 2]
```diff
@@ -879,7 +875,66 @@ template<class TPointType> class Triangle3D3
...
+        MathUtils<double>::UnitCrossProduct(normal, tangent_xi, tangent_eta);
+
+        // The center of the geometry
+        const auto& r_center = this->Center();
+
+        // We compute the distance, if it is not in the pane we
+        CoordinatesArrayType point_projected = rPoint;
+        const array_1d<double,3> vector_points = rPoint - r_center.Coordinates();
+        const double distance = inner_prod(vector_points, normal);
+        if (std::abs(distance) > std::numeric_limits<double>::epsilon()) { // Not in the plane, projecting
```
Comment (philbucher) [inline kratos/geometries/triangle_3d_3.h:896]: wait, doesn't this function assume that the point is in-plane?

I think this is a behavior-change?

**Comment by philbucher** [comment]
So I debugged a bit and my failing test is related to the tolerance (doing isInside on a corner)
Before this returned `true`, now it returns `false`

Therefore marking as behavior-change

Will keep investigaing

**Comment by philbucher** [comment]
screenshot from 2019-01-31 16-53-22

For the record:
It seems like the changes introduced some more numerical errors
Before most of the time the result of the computed local-coordinates was exactly 0.0, now it is ~1E-15
The tolerance was the numerical-limit epsilon, which works for one but not for the other case.

If I specify a larger tolerance my tests work again.

For me this is fine, what do you think @KratosMultiphysics/technical-committee ?

Also regarding the projection, which I think should not be done here, we have special functions for this

**Comment by loumalouomega** [comment]
The projection was proposed by @RiccardoRossi, it is the only way to do it in a non iterative way

**Comment by loumalouomega** [comment]
The thing is that if you assume the point is in the plane you should define
a tolerance, and until 1.0e-6 I did not work all the tests. So I remove it.

**Comment by jcotela** [comment]
Note that not checking if the point is on the plane or not makes IsInside checks give wrong results... However, I believe that the old implementation was already behaving like this. There should be an issue about that somewhere, but I could not find it. Maybe @pooyan-dadvand or @roigcarlo remember it better than me?

**Comment by loumalouomega** [comment]
> Note that not checking if the point is on the plane or not makes IsInside checks give wrong results... However, I believe that the old implementation was already behaving like this. There should be an issue about that somewhere, but I could not find it. Maybe @pooyan-dadvand or @roigcarlo remember it better than me?

The old implementation only chekced the residual of the local coordinates, so it could be outside the plane and still give a good result. I decided not to check the the distance to the plane because the tolerances are dangerous...(and it didn't passed the tests until certain value)

**Comment by pooyan-dadvand** [comment]
Too late but for me the best behaviour would be testing the in plane and in contrast to  @loumalouomega experience, for me the algorithm with distance is very practical.

**Comment by loumalouomega** [comment]
> Too late but for me the best behaviour would be testing the in plane and in contrast to @loumalouomega experience, for me the algorithm with distance is very practical.

Do you mean to check the distance to be below a certain threshold?

**Comment by pooyan-dadvand** [comment]
I expressed myself badly. The check for distance to plane is very practical (and I do it with a tolerance) because always after a calling to the previous version of IsInside I had to see if the point is on the plane. Doing it inside would improve the performance in my case because you can avoid the calculation of shape function if the point is not on the plane.

**Comment by loumalouomega** [comment]
> I expressed myself badly. The check for distance to plane is very practical (and I do it with a tolerance) because always after a calling to the previous version of IsInside I had to see if the point is on the plane. Doing it inside would improve the performance in my case because you can avoid the calculation of shape function if the point is not on the plane.

I will do an implementation for that

**Comment by loumalouomega** [comment]
@pooyan-dadvand I created core/tolerance-distance-local-coord-Triangle3D3N. I will run all the tests and I will see, then PR

**Pull request merged**